core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java

```diff
@@ -0,0 +1,171 @@
+import org.apache.spark.util.Utils;
+
+public class DefaultShuffleMapOutputWriter implements ShuffleMapOutputWriter {
+
+  private static final Logger log =
+    LoggerFactory.getLogger(DefaultShuffleMapOutputWriter.class);
+
+  private final int shuffleId;
+  private final int mapId;
+  private final int numPartitions;
+  private final ShuffleWriteMetricsReporter metrics;
+  private final IndexShuffleBlockResolver blockResolver;
+  private final long[] partitionLengths;
+  private ShufflePartitionWriter[] partitionWriters;
+  private final int bufferSize;
+  private int currPartitionId = 0;
+
+  private final File outputFile;
+  private final File outputTempFile;
+  private FileOutputStream outputFileStream;
+  private TimeTrackingOutputStream ts;
+  private BufferedOutputStream outputBufferedFileStream;
+
+  public DefaultShuffleMapOutputWriter(
+      int shuffleId,
+      int mapId,
+      int numPartitions,
+      ShuffleWriteMetricsReporter metrics,
+      IndexShuffleBlockResolver blockResolver,
+      SparkConf sparkConf) {
+    this.shuffleId = shuffleId;
+    this.mapId = mapId;
+    this.numPartitions = numPartitions;
+    this.metrics = metrics;
+    this.blockResolver = blockResolver;
+    this.bufferSize =
+      (int) (long) sparkConf.get(
+        package$.MODULE$.SHUFFLE_UNSAFE_FILE_OUTPUT_BUFFER_SIZE()) * 1024;
+    this.partitionLengths = new long[numPartitions];
+    this.partitionWriters = new ShufflePartitionWriter[numPartitions];
+    this.outputFile = blockResolver.getDataFile(shuffleId, mapId);
+    this.outputTempFile = Utils.tempFileWith(outputFile);
+  }
+
+  @Override
+  public ShufflePartitionWriter getNextPartitionWriter() throws IOException {
+    initStream();
+    ShufflePartitionWriter shufflePartitionWriter =
+      new DefaultShufflePartitionWriter(
+        new DefaultShuffleBlockOutputStreamImpl());
+    partitionWriters[currPartitionId++] = shufflePartitionWriter;
+    return shufflePartitionWriter;
+  }
+
+  @Override
+  public void commitAllPartitions() throws IOException {
+    for (int pId = 0; pId < numPartitions; pId ++) {
+      partitionLengths[pId] = partitionWriters[pId].getLength();
+    }
+    blockResolver.writeIndexFileAndCommit(shuffleId, mapId, partitionLengths, outputTempFile);
+    // TODO: Maybe monitor a little more intelligably then checking files?
+    if (!outputFile.getParentFile().isDirectory() && !outputFile.getParentFile().mkdirs()) {
+      throw new IOException(
+              String.format(
+                      "Failed to create shuffle file directory at %s.",
+                      outputFile.getParentFile().getAbsolutePath()));
+    }
+    if (!outputFile.isFile() && !outputFile.createNewFile()) {
+      throw new IOException(
+              String.format(
+                      "Failed to create empty shuffle file at %s.", outputFile.getAbsolutePath()));
+    }
+  }
+
+  @Override
+  public void abort(Throwable error) throws IOException {
+    cleanUp();
```

Comment:
When aborting, we don't want `cleanup` to throw an exception that halts the rest of this method - we want to always be able to attempt to delete the files.